core/templates/dev/head/components/question-directives/question-editor/question-editor.constants.ts

```diff
@@ -0,0 +1,19 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Constants for the questions editor directive.
+ */
+
+oppia.constant('INTERACTION_SPECS', GLOBALS.INTERACTION_SPECS);
```

Comment:
If INTERACTION_SPECS is used in both question-editor and question-player (and possibly elsewhere), shouldn't it be an app-level constant? Or perhaps not at the root, but somewhere shared? We should not be duplicating code between question-editor and question-player (or multiple files in general).

Also, what steps are you planning to take (in terms of presubmit checks) to ensure that every constant is declared in at most one place? That's something that can definitely happen if there are constants in multiple files.

Reply:
Actually, this is a special type of "constant". It uses the GLOBALS variable which is different for every page and hence needs to be declared at individual pages. I tried putting it at the and it expectedly gave errors.

Reply:
I think your last sentence is missing a word :)

Also I think @vojtechjelinek and @Jamesjay4199 might need to comment on this, this sounds like something that needs to be fixed (and not in this way).

Reply:
If you look at https://github.com/oppia/oppia/search?l=Python&q=INTERACTION_SPECS you can see that `INTERACTION_SPECS` is always set to `interaction_registry.Registry.get_all_specs()` so the value is always the same, the problem is that we are sending this only on some pages and if we were to move to app-level constant we would have to send it with every page.

Let's keep multiple constants for now and @Jamesjay4199 will refactor this in his third milestone.

Reply:
OK, two follow-up questions then.

(1) 

> It uses the GLOBALS variable which is different for every page and hence needs to be declared at individual pages. I tried putting it at the and it expectedly gave errors.

@YashJipkate can you elaborate on this (the "different for every page" part)? It contradicts with what @vojtechjelinek said, so I would like to understand the evidence for your statement.

(2) @vojtechjelinek I don't think we can support multiple repetitions of the same constant. Can we have a separate file containing just this constant that is imported by the different pages? That way the constant stays in one place, and @YashJipkate can put in checks to enforce the constants policy. There should be a TODO and an issue to delete this file and that issue should be assigned to @Jamesjay4199. Would that work?

Reply:
@seanlip I did not know the source of the variable (that it was coming from backend) and since it was giving out errors, so I thought maybe the constant was not a constant for every page. You can ignore that statement of mine.

Reply:
@seanlip (2) Sure that's okay

Reply:
Yeah, (2) seems okay @seanlip